526: Implement more accurate "We're missing some information" messages

NB28VT opened the pull request:
## Summary

### Overview

Implements the front end portion of a highly-requested fix to confusing and misleading error messages Users encounter when attempting to fill out Form 526. This has been a significant source of frustration for Veterans, as the message blocks the User from proceeding and prompts them to reach out to a Contact Center, but only provides a limited and in some cases completely inaccurate indication of what the problem is. For a detailed explanation of the back and front end problems with the exisitng implementation see report [here](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/disability/526ez/engineering_research/526_entry_error_discovery.md)

A recent [round of Medalia research](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/disability/526ez/research/Medallia%20Research/2023-10%20Research%20Findings.md) made this need even more clear.

It was only necessary to feature gate this on the back end (implemented [here](https://github.com/department-of-veterans-affairs/vets-api/pull/14858)). Will explain why below.

Here is a demo of the UX for the page before and after the changes:

https://github.com/department-of-veterans-affairs/vets-website/assets/8537428/ddda8c49-5a02-46c9-87ef-806545a01fe3

### Code Changes:
- Implements a new component, `Missing526Identifiers`, which will replace the various user-facing error message components in the [MissingServices file](https://github.com/department-of-veterans-affairs/vets-website/blob/dbex-62119-missing-id-messages-component/src/applications/disability-benefits/all-claims/containers/MissingServices.jsx).
- The missing data points we check for are BIRLS ID, Participant ID, Social Security Number, Date of Birth, and EDIPI.
- The new `Missing526Identifiers` component addresses this by checking exactly which data points are missing in some metada I added to the `claims` portion of the `user` profile in https://github.com/department-of-veterans-affairs/vets-api/pull/14858
- There was no need to feature gate this on the front end - the new metadata only appears if the feature gate is enabled in the backend, and I've written the front end to only check for the presence of that metadata when deciding whether to display the new message or not, so it's agnostic to the feature gate state for simplicity.


### Additional Required Summary Responses
- Which team do you work for, does your team own the maintenance of this component?: **I am on the DBex team, working specifically to improve submission success rates on the 526ez form. This work will only affect the 526ez flow, which is maintained by myself and my team.**
- (If using a flipper, what is the end date of the flipper being required/success criteria being targeted) **Unknown, my team is still new and working out what our release strategy will be, but I imagine after a couple weeks of having these improved messages live we will feel comfortable enabling them for all users.**


## Related issue(s)
- See epic and referenced reports mentioned above for more detail on how the problem was identified and addressed


## Testing done
- Unit tests were created for the new component and the form tests in `src/applications/disability-benefits/all-claims/tests/containers/Form526EZApp.unit.spec.jsx` were updated to test both the old and new messages with the feature flag.
- I needed to work with the Mock Data repo and a lot of trial and error to get into a state where I could test the new messages. Some data (`participant_id`, `edipi`) I could not figure out how to manipulate by altering the various XML and JSON files I had for test users locally.
- For testing end to end locally, I had to edit mock data I had for certain users and clear their cached data in redis so new data would appear; the process I used is well summarized by Sam Stuckey [in this doc ](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/disability/526ez/engineering_research/testing_with_mockdata_users.md)
- Test Plan: N/A I don't believe my team is planning on doing these for projects this small.

## Screenshots

See video demo above

## What areas of the site does it impact?

These messages appear at the beginning of the 526 flow when users are in a state where we are missing their data

## Acceptance criteria

### Quality Assurance & Testing

- [X] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [X] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [X] Linting warnings have been addressed
- [X] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### Error Handling

- [ ] Browser console contains no warnings or errors.
- [ ] Events are being sent to the appropriate logging solution
- [ ] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [ ] Did you login to a local build and verify all authenticated routes work as expected with a test user

### :warning: Team Sites (only applies to modifications made to the VA.gov header) :warning:

- [ ] The vets-website header does not contain any web-components
- [ ] I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#local-dev) to test the injected header scenario
- [ ] I reached out in the `#sitewide-public-websites` Slack channel for questions

## Requested Feedback

It's recommended to look at this commit by commit to understand the changes; I kept things like implementing Google Analytics tracking and adding some convenience utilities in their own discrete change.

I'm curious about the strategy I used for logging when these messages appear in Google Analytics. Since I know exactly what data is missing we have a great opportunity here to log the problem in more detail. I opted to dynamically build the `error-key` for Google Analytics like so:

```ruby
    'error-key': `missing_526_identifiers_${missingIdentifiers.join('_')}`,
```

.. so the key will appear like this for example:

`missing_526_identifiers_participantId_ssn_edipi`


However, I'm not sure how easily searchable this will be in Google Analytics because the combination of keys will vary. Any suggestions on maybe a better way to do this

NB28VT committed:
Implement 526 missing identifiers warning component

Creates a basic component and unit tests to tell the user they are missing certain identifiers we require before starting the Form 526 flow, and dynamically renders which identifiers are missing so they can share that information with a contact center

This is intended to replace all of the the inconsistent and inaccurate message components in src/applications/disability-benefits/all-claims/containers/MissingServices.jsx

NB28VT committed:
Refactor Missing526Identifiers containerm tests

Cleans up the formatting and rendering code for Missing526Identifiers to make it more logically organized. Addresses problems with tests that seemed to be working before; ended up having to use a regexp match on the container content for my message paragraph; using typical React Testing Library functions such as getByText is problematic and causes confusing failures because it doesn't support web components, which va-alert is (it's part of the VA component library)

NB28VT committed:
Add Google Analytics logging to message display

Adds a recordEvent call to log the appearance of the error message and dynamically includes which data was missing. May consider a better way to include the list of missing identifiers in the analytics log so it is easier to query on in the future.

Fix typo on participantId key in test and prop types

NB28VT committed:
Refactor Form526 to render new error messages

Replaces the existng Missing Services component renders in Form526EZApp.jsx with the new Missing526Identifiers component. This was done to standardize the messsaging during these situations and to provide the user with exactly which data is missing to share with the contact center.

Adds default claims arg to the test page in Form526EZApp.unit.spec.jsx. These styles of tests are now deprecated in our system, but I kept my new tests in the same style for consistency.

NB28VT committed:
Refactor list formatting for older browser compatibility

Adds a new utility function to our claims utilities for formatting a list as a human readable string with the correct commas and use of 'and'.

Replaces my initial implementation of this in the Missing526Identifiers endpoint which used the Intl.ListFormat function:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat

As this is a relatively new feature of JavaScript not supported by older browsers (I was told Safari 9 for example, which many veterans are still on, frequently raises errors when we use these features), it's better to use the more vanilla, classic JavaScript approach

NB28VT committed:
Feature gate new and old messaging behavior

Adds back the existing messages behavior I had removed in an earlier commit and wraps the new and old behavior in a feature flag.

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

NB28VT commented:
Additional project and UX-related notes for @austinjprice @juliepedtke @nnicholas7

* The generic services message will still appear if there is a problem adding the user to MVI (recall [we decided to punt on this](https://dsva.slack.com/archives/C05MPTGMX0E/p1692647266596259), which I agree was the right move)
* Based on what I know about what happens under the hood, that generic MVI message will probably be more common still than the improved ones. But at least more detailed messages will be clearer.
* See my notes above under the "Requested Feedback" section regarding Google Analytics. I haven't been into DOMO yet so I'm not sure how the way I'm logging the error keys will work from a practical standpoint as we track how often these missing keys appear. Happy to chat about this on a call if it's easier

NB28VT commented:
This is my first PR in this repo - how seriously are the [Code Climate](https://codeclimate.com/github/department-of-veterans-affairs/vets-website/pull/25428) suggestions here enforced? Some of these already existed and a couple seems a little excessive (25 lines vs 29 for a function).

Does the reviewer have any opinions on which of these I need to implement to get approval here?

zack commented on src/applications/disability-benefits/all-claims/containers/Missing526Identifiers.jsx:
```diff
@@ -0,0 +1,104 @@
+export const Missing526Identifiers = ({ title, form526RequiredIdentifers }) => {
+  const alertContent = displayContent(title, form526RequiredIdentifers);
+
+  return <Alert title={title} content={alertContent} />;
```
🌴 By no means necessary and just a light suggestion, but this seems like a pretty good use case for `{children}`:

```js
return (
  <Alert title={title}>
    {alertContent}
  <Alert/>;
);
```

**Alert.js**:
```js
const Alert = ({ children, title }) => (
  <div className="vads-l-grid-container vads-u-padding-left--0 vads-u-padding-bottom--5">
    <div className="usa-content">
      <h1>{title}</h1>
      <va-alert status="error">{children}</va-alert>
    </div>
  </div>
);
```

NB28VT replied:
I like it - https://github.com/department-of-veterans-affairs/vets-website/pull/25428/commits/fe3bf2792ddb1b9f6f126484812c146868d68261

zack commented on src/applications/disability-benefits/all-claims/containers/Missing526Identifiers.jsx:
```diff
@@ -0,0 +1,104 @@
+    form526RequiredIdentifers,
+  );
+
+  recordEvent({
```
What does this do? I'm looking at `record-event.js` and I can't figure out what it's taking (seemingly) styling type information for. 😅

NB28VT replied:
I can't find a VA doc for this but my understanding based on limited knowledge of Google Analytics and the other calls elsewhere in the codebase is that this metadata is intended to log the type of UI elements that were present when an error was logged, presumably for metrics related to accessibility.

zack commented on src/applications/disability-benefits/all-claims/tests/containers/Missing526Identifiers.unit.spec.jsx:
```diff
@@ -0,0 +1,101 @@
+      },
+    };
+
+    it("should render a message noting the missing identifiers with 'and' between them", () => {
```
Nit: should this be something like "noting the missing identifies in a comma separated list"?

NB28VT replied:
Yeah this language did feel clunky when I wrote it, this was a good suggestion and I cleaned up another one on this page in https://github.com/department-of-veterans-affairs/vets-website/pull/25428/commits/a67744174b4444bac1d688e31d562da4efd21a29

zack reviewed (dismissed):
Pretty new to this part of the codebase but it looks good! Just one change I'd strongly suggest (`title = ''`), and a few non-blocking question and nits. 💪🍔👍

NB28VT commented on src/applications/disability-benefits/all-claims/utils/index.jsx:
```diff
@@ -666,6 +685,9 @@ export const showSeparationLocation = formData => {
 
 export const show526Wizard = state => toggleValues(state).show526Wizard;
 
+export const form526RequiredIdentifiersInUserObject = state =>
```
Yeah, it's not very intuitive is it? Hhaha. This is the VA convention for [implementing a Redux selector for toggling feature gates](https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-7.Usetheselectorhelpertobuildaselectorforyourfeaturetoggle). There are a couple selectors on this page maybe I'll just add a comment

NB28VT replied:
Decided not to, these appear in several places throughout the codebase and adding a comment just for my new one didn't really seem consistent.

NB28VT committed:
Add empty string safeguard to title formatter

The default argument to this titleLowerCase function is an empty string, however the destructuring below it will leave the firstChar undefined and would error if we tried to lower case it

NB28VT committed:
Refactor Alert to use React children pattern

NB28VT committed:
Improve language in Missing526Identifiers spec

NB28VT commented on src/applications/disability-benefits/all-claims/containers/Missing526Identifiers.jsx:
```diff
@@ -0,0 +1,104 @@
+// This is needed because we reference it within a sentance
+const titleLowerCase = (title = '') => {
+  const [firstChar, ...rest] = title;
+  return `${firstChar.toLowerCase()}${rest.join('')}`;
```
Wow great catch - fixed in https://github.com/department-of-veterans-affairs/vets-website/pull/25428/commits/472b85417a3ae74923b3b2964cc2391f090faee7

SamStuckey commented on src/applications/disability-benefits/all-claims/Form526EZApp.jsx:
```diff
@@ -64,6 +65,13 @@ export const hasRequiredId = profile =>
+// Will replace hasRequiredServices, hasRequiredId and hasRequiredDob when form526RequiredIdentifiersInUserObject feature flag is removed
+const missingRequiredIdentifiers = profile => {
+  return Object.values(profile.claims.form526RequiredIdentifierPresence).some(
+    idPresence => idPresence === false,
```
will this value always be `false`?  or is there a chance it will be `null`?  what would that do to this code?

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

NB28VT committed:
Remove feature flag toggles

Removes the front end feature flags I initially implemented here alongside the back end changes in https://github.com/department-of-veterans-affairs/vets-api/pull/14858. I decided it would simplify the code and release rollout to simply check if the new missing identifiers metadata was present (meaning the back end feature toggle was on), and make the component rendering dependent on that instead of checking a flag that is unecessary in this case.

NB28VT committed:
Revert accidental removal of important docstring from existing code

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

Mottie commented on src/applications/disability-benefits/all-claims/containers/Missing526Identifiers.jsx:
```diff
@@ -0,0 +1,108 @@
+
+  return (
+    <>
+      <h2 className="vads-u-display--inline-block vads-u-font-size--h3 vads-u-margin-top--0">
+        We need more information for your application
+      </h2>
```
Since this is inside a `va-alert`, this `h2` should include a `slot="headline"` property

Mottie commented on src/applications/disability-benefits/all-claims/utils/index.jsx:
```diff
@@ -206,6 +206,25 @@ export const capitalizeEachWord = name => {
+ * @param {string[]} list
+ * @returns {string} formatted string
+ */
+export const formatList = list => {
+  if (list.length === 1) {
+    return list[0];
+  }
+  if (list.length === 2) {
+    return list.join(' and ');
+  }
+  const lastItem = list.pop();
+  return `${list.join(', ')}, and ${lastItem}`;
+};
```
There is a [`readableList` utility](https://github.com/department-of-veterans-affairs/vets-website/blob/main/src/platform/forms-system/src/js/utilities/data/readableList.js) that does this same thing

Mottie reviewed (dismissed):
LGTM. It'd be nice if this was included as a platform component where you can set which bits of info are required; but that isn't on you to do!

NB28VT committed:
Add slot attr to h2 in VA alert

NB28VT committed:
Use existing utility for formatting lists

I wasn't aware of this readable list utility when I made my own, so deleting my implementation and relying on this existing helper.

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

jerekshoe commented on src/applications/disability-benefits/all-claims/containers/Missing526Identifiers.jsx:
```diff
@@ -0,0 +1,111 @@
+import readableList from 'platform/forms-system/src/js/utilities/data/readableList';
+
+// Downcases the first char in the capitalized form title passed to this component
+// This is needed because we reference it within a sentance
```
Small typo: sentance should be sentence

NB28VT committed:
Fix docstring typo

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

freeheeling commented on src/applications/disability-benefits/all-claims/tests/containers/Form526EZApp.unit.spec.jsx:
```diff
@@ -147,60 +149,142 @@ describe('Form 526EZ Entry Page', () => {
-      verified: true,
-      services: [],
+  describe('Missing user identifiers', () => {
+    // claims.form526RequiredIdentifierPresence is included in the profile when the form_526_required_identifiers_in_user_object feature flag is enabled on the back end
```
Because the following test pertains to a claims object **without** `form526RequiredIdentifierPresence`, I would suggest modifying the wording to the effect of "...is **not** included..." and "...is **not** enabled..."

freeheeling reviewed (dismissed):
Looks good. Just one minor suggestion.

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

NB28VT commented:
The Code Climate suggestions are good guidelines but in this case I don't think they're blocking.

christinec-fftc commented on src/applications/disability-benefits/all-claims/containers/Missing526Identifiers.jsx:
```diff
@@ -0,0 +1,111 @@
+  <div className="vads-l-grid-container vads-u-padding-left--0 vads-u-padding-bottom--5">
+    <div className="usa-content">
+      <h1>{title}</h1>
+      <va-alert status="error">{children}</va-alert>
```
Have you considered adding the `uswds` prop for ui/ux consistency? (#27764 for reference)

NB28VT replied:
Ah wasn't aware of that one good to know thanks! Fixed in https://github.com/department-of-veterans-affairs/vets-website/pull/25428/commits/c69b8afdf2d2001806d5874edac06456f5c85325

NB28VT committed:
PR Feedback: Add v3 uswds styling tag to alert message

Adds the uswds tag to the alert poriton of the new More Info Needed error message for styling consistency

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

NB28VT commented:
@christinec-fftc just merged master again here, the CI tasks with test stability seem to be failing intermittently, not sure why, if there are still failures when you get around to re-reviewing this and you have some insight feel free to let me know thanks! I'm not really sure what some of those are actually doing

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

NB28VT commented:
> @christinec-fftc just merged master again here, the CI tasks with test stability seem to be failing intermittently, not sure why, if there are still failures when you get around to re-reviewing this and you have some insight feel free to let me know thanks! I'm not really sure what some of those are actually doing

Disregard - this is related to a problem with `main` that platform is aware of; Robin ran into this as well and opened a ticket

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component

NB28VT committed:
Merge branch 'main' into dbex-62119-missing-id-messages-component
